File: src/GBA/PPU.cs
namespace gbaemu.GBA {
    public class PPU : BusPeripheral {
        public GBA GBA {get; private set;}

        private byte[] palette;
        private byte[] vram;
        private byte[] oam;

        public delegate void OnFrame(uint[] color);
        public event OnFrame OnFrameEvent;
        public delegate void OnVblank();
        public event OnVblank OnVblankEvent;
        public delegate void OnHblank();
        public event OnHblank OnHblankEvent;

        private uint[] frameBuffer;
        private uint currentRow;
        private uint currentCol;
        private uint currentCycle;

        public PPU(GBA gba) {
            GBA = gba;
            frameBuffer = new uint[240 * 160];
            palette = new byte[1024];
            vram = new byte[98304];
            oam = new byte[1024];
        }

        public override byte Read8(uint addr) {
            switch((addr & 0x0f000000) >> 24) {
                case 0x5:
                    return palette[addr & 0x000003ff];

                case 0x6:
                    addr &= 0x0001ffff;

                    if(addr >= 0x00018000) {
                        addr &= 0x00017fff;
                    }

                    return vram[addr];

                case 0x7:
                    return oam[addr & 0x000003ff];
            }

            return 0xff;
        }

        public override void Write8(uint addr, byte val) {
            Write16(addr, (ushort)((val << 8) | val));
        }

        public new void Write16(uint addr, ushort val) {
            addr &= 0xfffffffe;

            switch((addr & 0x0f000000) >> 24) {
                case 0x5:
                    addr &= 0x000003fe;
                    palette[addr] = (byte)val;
                    palette[addr + 1] = (byte)(val >> 8);
                    break;
                    
                case 0x6:
                    addr &= 0x0001ffff;

                    if(addr >= 0x00018000) {
                        addr &= 0x00017fff;
                    }
                    
                    vram[addr] = (byte)val;
                    vram[addr + 1] = (byte)(val >> 8);
                    break;

                case 0x7:
                    addr &= 0x000003fe;
                    oam[addr] = (byte)val;
                    oam[addr + 1] = (byte)(val >> 8);
                    break;
            }
        }

        private uint ColorToRGB(ushort color) {
            uint blue = (uint)(color & 0x7c00) >> 10;
            uint green = (uint)(color & 0x03e0) >> 5;
            uint red = (uint)color & 0x001f;

            return 0xff000000 | (red << 3) | (green << 11) | (blue << 19);
        }

        private ushort GetPaletteColor(byte paletteIndex) {
            return Read16(0x05000000U | (uint)(paletteIndex << 1));
        }

        private void SortLayers(out int[] layers) {
            int[] intLayers = new int[4] {3, 2, 1, 0};
            int[] priorities = new int[4];

            for(int i = 0; i < 4; i++) {
                priorities[i] = GBA.Io.GetRegister(0x04000008 + ((uint)i << 1)).Value & 0x0003;
            }
            
            for(int i = 0; i < 3; i++) {
                int layerPriority = priorities[i];
                int maxPriority = layerPriority;
                int maxPriorityIndex = i;

                for(int j = i + 1; j < 4; j++) {
                    int currentLayerPriority = priorities[j];

                    if(currentLayerPriority > maxPriority) {
                        maxPriorityIndex = j;
                        maxPriority = currentLayerPriority;
                    }
                }

                int exchange = intLayers[i];
                intLayers[i] = intLayers[maxPriorityIndex];
                intLayers[maxPriorityIndex] = exchange;
            }

            layers = intLayers;
        }

        private void DrawLayer(int layer) {
            IO.Register bgcnt = GBA.Io.GetRegister(0x04000008 + ((uint)layer << 1));

            uint hofs = GBA.Io.GetRegister(0x04000010 + ((uint)layer << 2)).Value;
            uint vofs = GBA.Io.GetRegister(0x04000012 + ((uint)layer << 2)).Value;
            uint mapBase = (uint)(bgcnt.Value & 0x1f00) << 3;
            uint tileBase = (uint)(bgcnt.Value & 0x000c) << 12;
            uint yLayer = currentRow + vofs;

            uint mapOffsetY = 0x00000000;

            if(BitUtils.BitTest16(bgcnt.Value, 15)) {
                yLayer &= 0x000001ff;

                if(((bgcnt.Value & 0xc000) == 0xc000) && (yLayer >= 0x100)) {
                    mapOffsetY = 0x00001000;
                }
            } else {
                yLayer &= 0x000000ff;
            }

            uint yChunk = yLayer & 0x000000ff;

            uint yMap = yChunk >> 3;
            uint yTile = yChunk & 7;

            for(uint x = 0; x < 240; x++) {
                uint xLayer = x + hofs;

                uint mapOffsetX = 0x00000000;

                if(BitUtils.BitTest16(bgcnt.Value, 14)) {
                    xLayer &= 0x000001ff;

                    if(BitUtils.BitTest16(bgcnt.Value, 14) && (xLayer >= 0x100)) {
                        mapOffsetX = 0x00000800;
                    }
                } else {
                    xLayer &= 0x000000ff;
                }

                uint xChunk = xLayer & 0x000000ff;

                uint xMap = xChunk >> 3;
                uint xTile = xChunk & 7;

                uint mapOffset = mapOffsetX + mapOffsetY;

                uint mapAddress = mapOffset + (yMap << 6) + (xMap << 1);
                uint mapValue = vram[mapAddress] | ((uint)vram[mapAddress + 1] << 8);

                uint tileNumber = mapValue & 0x03ff;
                bool flipHorizontal = BitUtils.BitTest32(mapValue, 10);
                bool flipVertical = BitUtils.BitTest32(mapValue, 11);

                uint yTileReal = flipVertical ? 7 - yTile : yTile;
                uint xTileReal = flipHorizontal ? 7 - xTile : xTile;

                uint colorAddress;

                if(BitUtils.BitTest16(bgcnt.Value, 7)) {
                    uint tileDataAddress = (tileNumber << 6) | (yTileReal << 3) | xTileReal;
                    uint tileValue = vram[tileBase + tileDataAddress];
                    colorAddress = tileValue << 1;
                } else {
                    uint tileDataAddress = (tileNumber << 5) | (yTileReal << 2) | (xTileReal >> 1);
                    
                    uint tileValue = vram[tileBase + tileDataAddress];

                    if(BitUtils.BitTest32(xTileReal, 0) ^ flipHorizontal) {
                        tileValue >>= 4;
                    } else {
                        tileValue &= 0x0f;
                    }

                    uint paletteNumber = mapValue >> 12;

                    colorAddress = (paletteNumber << 5) | (tileValue << 1);
                }

                ushort pixelColor = (ushort)(palette[colorAddress] | (palette[colorAddress + 1] << 8));

                frameBuffer[currentRow * 240 + x] = ColorToRGB(pixelColor);
            }
        }

        private void DrawMode0() {
            int[] layers;

            SortLayers(out layers);
            
            IO.Register dispcnt = GBA.Io.GetRegister(0x04000000);

            for(int i = 0; i < 4; i++) {
                if((dispcnt.Value & (1 << (8 + i))) != 0) {
                    DrawLayer(i);
                }
            }
        }

        private void DrawMode1() {
            DrawMode0();
        }

        private void DrawMode2() {
            DrawMode0();
        }

        private void DrawMode3() {
            for(int x = 0; x < 240; x++) {
                frameBuffer[currentRow * 240 + x] = ColorToRGB(Read16(0x06000000 | (uint)((currentRow * 240 + x) << 1)));
            }
        }

        private void DrawMode4() {
            IO.Register dispcnt = GBA.Io.GetRegister(0x04000000);

            uint offset = BitUtils.BitTest16(dispcnt.Value, 4) ? 0x0000a000U : 0x00000000U;

            for(int x = 0; x < 240; x++) {
                frameBuffer[currentRow * 240 + x] = ColorToRGB(GetPaletteColor(vram[currentRow * 240 + x + offset]));
            }
        }

        private void DrawMode5() {
            IO.Register dispcnt = GBA.Io.GetRegister(0x04000000);

            uint offset = BitUtils.BitTest16(dispcnt.Value, 4) ? 0x00005000U : 0x00000000U;

            uint currentRow = this.currentRow - 16;

            if(currentRow < 128) {
                for(int x = 0; x < 160; x++) {
                    frameBuffer[currentRow * 240 + x + 40] = ColorToRGB(Read16(0x06000000 | (uint)((currentRow * 160 + x + offset) << 1)));
                }
            }
        }

        private void DrawLine() {
            IO.Register dispcnt = GBA.Io.GetRegister(0x04000000);

            switch(dispcnt.Value & 0x07) {
                case 0: DrawMode0(); break;
                case 1: DrawMode1(); break;
                case 2: DrawMode2(); break;
                case 3: DrawMode3(); break;
                case 4: DrawMode4(); break;
                case 5: DrawMode5(); break;
                default: break;
            }
        }

        public void Cycle() {
            if((currentCycle & 0x03) == 0x03) {
                IO.Register dispstat = GBA.Io.GetRegister(0x04000004);

                currentCol++;

                if(currentCol == 308) {
                    currentCol = 0;
                    currentRow++;

                    dispstat.Value = BitUtils.BitClear16(dispstat.Value, 1);

                    if((dispstat.Value >> 8) == currentRow) {
                        dispstat.Value = BitUtils.BitSet16(dispstat.Value, 2);

                        if(BitUtils.BitTest16(dispstat.Value, 5)) {
                            GBA.SetInterruptFlag(1 << 2);
                        }
                    } else {
                        dispstat.Value = BitUtils.BitClear16(dispstat.Value, 2);
                    }

                    if(currentRow == 228) {
                        currentRow = 0;
                        currentCycle = 0;
                        dispstat.Value = BitUtils.BitClear16(dispstat.Value, 0);
                    } else if(currentRow == 160) {
                        dispstat.Value = BitUtils.BitSet16(dispstat.Value, 0);

                        if(BitUtils.BitTest16(dispstat.Value, 3)) {
                            GBA.SetInterruptFlag(1 << 0);
                            
                            if(OnVblankEvent != null) {
                                OnVblankEvent.Invoke();
                            }
                        }

                        OnFrameEvent(frameBuffer);
                    }

                    GBA.Io.GetRegister(0x04000006).Value = (ushort)currentRow;
                } else if(currentCol == 240) {
                    dispstat.Value = BitUtils.BitSet16(dispstat.Value, 1);

                    if(BitUtils.BitTest16(dispstat.Value, 4)) {
                        GBA.SetInterruptFlag(1 << 1);

                        if(OnHblankEvent != null) {
                            OnHblankEvent.Invoke();
                        }
                    }

                    if(currentRow < 160) {
                        DrawLine();
                    }
                }
            }

            currentCycle++;
        }
    }
}

File: src/core/ppu.c
#include <stdint.h>
#include <string.h>

#include "util.h"
#include "core/defines.h"
#include "core/dma.h"
#include "core/gba.h"
#include "core/io.h"
#include "frontend/frontend.h"

uint8_t gba_ppu_palette[GBA_PALETTE_SIZE];
uint8_t gba_ppu_vram[GBA_VRAM_SIZE];
uint8_t gba_ppu_oam[GBA_OAM_SIZE];

uint32_t gba_ppu_frameBuffer[GBA_SCREEN_WIDTH * GBA_SCREEN_HEIGHT];
uint_least32_t gba_ppu_currentRow;
uint_least32_t gba_ppu_currentColumn;
uint_least32_t gba_ppu_currentCycle;
unsigned int gba_ppu_layers[4];

void gba_ppu_reset();
uint8_t gba_ppu_palette_read8(uint32_t address);
uint16_t gba_ppu_palette_read16(uint32_t address);
uint32_t gba_ppu_palette_read32(uint32_t address);
void gba_ppu_palette_write8(uint32_t address, uint8_t value);
void gba_ppu_palette_write16(uint32_t address, uint16_t value);
void gba_ppu_palette_write32(uint32_t address, uint32_t value);
uint8_t gba_ppu_vram_read8(uint32_t address);
uint16_t gba_ppu_vram_read16(uint32_t address);
uint32_t gba_ppu_vram_read32(uint32_t address);
void gba_ppu_vram_write8(uint32_t address, uint8_t value);
void gba_ppu_vram_write16(uint32_t address, uint16_t value);
void gba_ppu_vram_write32(uint32_t address, uint32_t value);
uint8_t gba_ppu_oam_read8(uint32_t address);
uint16_t gba_ppu_oam_read16(uint32_t address);
uint32_t gba_ppu_oam_read32(uint32_t address);
void gba_ppu_oam_write8(uint32_t address, uint8_t value);
void gba_ppu_oam_write16(uint32_t address, uint16_t value);
void gba_ppu_oam_write32(uint32_t address, uint32_t value);
static inline uint32_t gba_ppu_colorToRgb(uint16_t color);
static inline uint16_t gba_ppu_getPaletteColor(uint8_t index);
static inline void gba_ppu_sortLayers();
static inline void gba_ppu_drawLayer(int layer);
static inline void gba_ppu_drawMode0();
static inline void gba_ppu_drawMode1();
static inline void gba_ppu_drawMode2();
static inline void gba_ppu_drawMode3();
static inline void gba_ppu_drawMode4();
static inline void gba_ppu_drawMode5();
static inline void gba_ppu_drawLine();
static inline void gba_ppu_onVblank();
static inline void gba_ppu_onHblank();

void gba_ppu_reset() {
    memset(gba_ppu_palette, 0, GBA_PALETTE_SIZE);
    memset(gba_ppu_vram, 0, GBA_VRAM_SIZE);
    memset(gba_ppu_oam, 0, GBA_OAM_SIZE);
}

void gba_ppu_cycle() {
    if((gba_ppu_currentCycle & 0x03) == 0x03) {
        gba_io_register_t *dispstat = gba_io_getRegister(0x04000004);

        if(gba_ppu_currentColumn == 308) {
            gba_ppu_currentColumn = 0;
            gba_ppu_currentRow++;

            dispstat->value &= ~(1 << 1);

            if((dispstat->value >> 8) == gba_ppu_currentRow) {
                dispstat->value |= (1 << 2);

                if(dispstat->value & (1 << 5)) {
                    gba_setInterruptFlag(1 << 2);
                }
            } else {
                dispstat->value &= ~(1 << 2);
            }

            if(gba_ppu_currentRow == 228) {
                gba_ppu_currentRow = 0;
                gba_ppu_currentCycle = 0;
                dispstat->value &= ~(1 << 0);
            } else if(gba_ppu_currentRow == GBA_SCREEN_HEIGHT) {
                dispstat->value |= (1 << 0);

                if(dispstat->value & (1 << 3)) {
                    gba_setInterruptFlag(1 << 0);
                }
                
                gba_ppu_onVblank();
            }

            gba_io_getRegister(0x04000006)->value = gba_ppu_currentRow;
        } else if(gba_ppu_currentColumn == GBA_SCREEN_WIDTH) {
            dispstat->value |= (1 << 1);

            if(dispstat->value & (1 << 4)) {
                gba_setInterruptFlag(1 << 1);
                gba_dma_onHblank();
            }

            if(gba_ppu_currentRow < GBA_SCREEN_HEIGHT) {
                gba_ppu_drawLine();
            }
        }

        gba_ppu_currentColumn++;
    }

    gba_ppu_currentCycle++;
}

uint8_t gba_ppu_palette_read8(uint32_t address) {
    return gba_ppu_palette[address & 0x000003ff];
}

uint16_t gba_ppu_palette_read16(uint32_t address) {
    return ACCESS_16(gba_ppu_palette, address & 0x000003fe);
}

uint32_t gba_ppu_palette_read32(uint32_t address) {
    return ACCESS_32(gba_ppu_palette, address & 0x000003fc);
}

void gba_ppu_palette_write8(uint32_t address, uint8_t value) {
    gba_ppu_palette_write16(address, (uint16_t)((value << 8) | value));
}

void gba_ppu_palette_write16(uint32_t address, uint16_t value) {
    ACCESS_16(gba_ppu_palette, address & 0x000003fe) = value;
}

void gba_ppu_palette_write32(uint32_t address, uint32_t value) {
    ACCESS_32(gba_ppu_palette, address & 0x000003fc) = value;
}

uint8_t gba_ppu_vram_read8(uint32_t address) {
    if(address & 0x00010000) {
        address &= 0x00017fff;
    } else {
        address &= 0x0001ffff;
    }

    return gba_ppu_vram[address];
}

uint16_t gba_ppu_vram_read16(uint32_t address) {
    if(address & 0x00010000) {
        address &= 0x00017ffe;
    } else {
        address &= 0x0001fffe;
    }

    return ACCESS_16(gba_ppu_vram, address);
}

uint32_t gba_ppu_vram_read32(uint32_t address) {
    if(address & 0x00010000) {
        address &= 0x00017ffc;
    } else {
        address &= 0x0001fffc;
    }

    return ACCESS_32(gba_ppu_vram, address);
}

void gba_ppu_vram_write8(uint32_t address, uint8_t value) {
    gba_ppu_vram_write16(address, (uint16_t)((value << 8) | value));
}

void gba_ppu_vram_write16(uint32_t address, uint16_t value) {
    if(address & 0x00010000) {
        address &= 0x00017ffe;
    } else {
        address &= 0x0001fffe;
    }

    ACCESS_16(gba_ppu_vram, address) = value;
}

void gba_ppu_vram_write32(uint32_t address, uint32_t value) {
    if(address & 0x00010000) {
        address &= 0x00017ffc;
    } else {
        address &= 0x0001fffc;
    }

    ACCESS_32(gba_ppu_vram, address) = value;
}

uint8_t gba_ppu_oam_read8(uint32_t address) {
    return gba_ppu_oam[address & 0x000003ff];
}

uint16_t gba_ppu_oam_read16(uint32_t address) {
    return ACCESS_16(gba_ppu_oam, address & 0x000003fe);
}

uint32_t gba_ppu_oam_read32(uint32_t address) {
    return ACCESS_32(gba_ppu_oam, address & 0x000003fc);
}

void gba_ppu_oam_write8(uint32_t address, uint8_t value) {
    gba_ppu_oam_write16(address, (uint16_t)((value << 8) | value));
}

void gba_ppu_oam_write16(uint32_t address, uint16_t value) {
    ACCESS_16(gba_ppu_oam, address & 0x000003fe) = value;
}

void gba_ppu_oam_write32(uint32_t address, uint32_t value) {
    ACCESS_32(gba_ppu_oam, address & 0x000003fc) = value;
}

static inline uint32_t gba_ppu_colorToRgb(uint16_t color) {
    uint32_t blue = (color & 0x7c00) >> 7;
    uint32_t green = (color & 0x03e0) >> 2;
    uint32_t red = color & 0x001f;

    red <<= 3;
    
    red |= red >> 5;
    green |= green >> 5;
    blue |= blue >> 5;

    return 0xff000000 | red | (green << 8) | (blue << 16);
}

static inline uint16_t gba_ppu_getPaletteColor(uint8_t index) {
    return gba_ppu_palette_read16(0x05000000 | (index << 1));
}

static inline void gba_ppu_sortLayers() {
    gba_ppu_layers[0] = 3;
    gba_ppu_layers[1] = 2;
    gba_ppu_layers[2] = 1;
    gba_ppu_layers[3] = 0;

    int priorities[4];

    for(int i = 0; i < 4; i++) {
        priorities[i] = gba_io_getRegister(0x04000008 + (i << 1))->value & 0x0003;
    }

    for(int i = 0; i < 3; i++) {
        int layerPriority = priorities[i];
        int maxPriority = layerPriority;
        int maxPriorityIndex = i;

        for(int j = i + 1; j < 4; j++) {
            int currentLayerPriority = priorities[j];

            if(currentLayerPriority > maxPriority) {
                maxPriority = currentLayerPriority;
                maxPriorityIndex = j;
            }
        }

        unsigned int exchange = gba_ppu_layers[i];
        gba_ppu_layers[i] = gba_ppu_layers[maxPriorityIndex];
        gba_ppu_layers[maxPriorityIndex] = exchange;
    }
}

static inline void gba_ppu_drawLayer(int layer) {
    gba_io_register_t *bgcnt = gba_io_getRegister(0x04000008 + (layer << 1));

    unsigned int hofs = gba_io_getRegister(0x04000010 + (layer << 2))->value;
    unsigned int vofs = gba_io_getRegister(0x04000012 + (layer << 2))->value;
    uint32_t mapBase = (bgcnt->value & 0x1f00) << 3;
    uint32_t tileBase = (bgcnt->value & 0x000c) << 12;
    unsigned int yLayer = gba_ppu_currentRow + vofs;

    uint32_t mapOffsetY = 0x00000000;

    if(bgcnt->value & (1 << 15)) {
        yLayer &= 0x000001ff;

        if(((bgcnt->value & 0xc000) == 0xc000) && (yLayer >= 0x100)) {
            mapOffsetY = 0x00001000;
        }
    } else {
        yLayer &= 0x000000ff;
    }

    unsigned int yChunk = yLayer & 0x000000ff;
    unsigned int yMap = yChunk >> 3;
    unsigned int yTile = yChunk & 7;

    for(unsigned int x = 0; x < GBA_SCREEN_WIDTH; x++) {
        unsigned int xLayer = x + hofs;
        uint32_t mapOffsetX = 0x00000000;

        if(bgcnt->value & (1 << 14)) {
            xLayer &= 0x000001ff;

            if((bgcnt->value & (1 << 14)) && (xLayer >= 0x100)) {
                mapOffsetX = 0x00000800;
            }
        } else {
            xLayer &= 0x000000ff;
        }

        unsigned int xChunk = xLayer & 0x000000ff;
        unsigned int xMap = xChunk >> 3;
        unsigned int xTile = xChunk & 0x07;
        uint32_t mapOffset = mapBase + mapOffsetX + mapOffsetY;
        uint32_t mapAddress = mapOffset + (yMap << 6) + (xMap << 1);
        uint16_t mapValue = gba_ppu_vram[mapAddress] | (gba_ppu_vram[mapAddress + 1] << 8);
        uint16_t tileNumber = mapValue & 0x03ff;
        bool flipHorizontal = (mapValue & (1 << 10)) != 0;
        bool flipVertical = (mapValue & (1 << 11)) != 0;

        unsigned int yTileReal = flipVertical ? 7 - yTile : yTile;
        unsigned int xTileReal = flipHorizontal ? 7 - xTile : xTile;

        uint32_t colorAddress;

        if(bgcnt->value & (1 << 7)) {
            uint32_t tileDataAddress = (tileNumber << 6) | (yTileReal << 3) | xTileReal;
            uint8_t tileValue = gba_ppu_vram[tileBase + tileDataAddress];
            colorAddress = tileValue << 1;
        } else {
            uint32_t tileDataAddress = (tileNumber << 5) | (yTileReal << 2) | (xTileReal >> 1);
            uint8_t tileValue = gba_ppu_vram[tileBase + tileDataAddress];

            if(!((xTileReal & (1 << 0)) ^ flipHorizontal)) {
                tileValue >>= 4;
            } else {
                tileValue &= 0x0f;
            }

            unsigned int paletteNumber = mapValue >> 12;

            colorAddress = (paletteNumber << 5) | (tileValue << 1);
        }

        uint16_t pixelColor = gba_ppu_palette[colorAddress] | (gba_ppu_palette[colorAddress + 1] << 8);
        gba_ppu_frameBuffer[gba_ppu_currentRow * GBA_SCREEN_WIDTH + x] = gba_ppu_colorToRgb(pixelColor);
    }
}

static inline void gba_ppu_drawMode0() {
    gba_ppu_sortLayers();

    gba_io_register_t *dispcnt = gba_io_getRegister(0x04000000);

    for(int i = 0; i < 4; i++) {
        if(dispcnt->value & (1 << (8 + i))) {
            gba_ppu_drawLayer(i);
        }
    }
}

static inline void gba_ppu_drawMode1() {
    gba_ppu_drawMode0();
}

static inline void gba_ppu_drawMode2() {
    gba_ppu_drawMode0();
}

static inline void gba_ppu_drawMode3() {
    for(int x = 0; x < GBA_SCREEN_WIDTH; x++) {
        gba_ppu_frameBuffer[gba_ppu_currentRow * GBA_SCREEN_WIDTH + x] = gba_ppu_colorToRgb(gba_ppu_vram_read16(0x06000000 | ((gba_ppu_currentRow * GBA_SCREEN_WIDTH + x) << 1)));
    }
}

static inline void gba_ppu_drawMode4() {
    gba_io_register_t *dispcnt = gba_io_getRegister(0x04000000);

    uint32_t offset = (dispcnt->value & (1 << 4)) ? 0x0000a000 : 0x00000000;

    for(int x = 0; x < GBA_SCREEN_WIDTH; x++) {
        gba_ppu_frameBuffer[gba_ppu_currentRow * GBA_SCREEN_WIDTH + x] = gba_ppu_colorToRgb(gba_ppu_getPaletteColor(gba_ppu_vram[gba_ppu_currentRow * GBA_SCREEN_WIDTH + x + offset]));
    }
}

static inline void gba_ppu_drawMode5() {
    gba_io_register_t *dispcnt = gba_io_getRegister(0x04000000);

    uint32_t offset = (dispcnt->value & (1 << 4)) ? 0x00005000 : 0x00000000;

    unsigned int currentRow = gba_ppu_currentRow - 16;

    if(currentRow < 128) {
        for(int x = 0; x < 160; x++) {
            gba_ppu_frameBuffer[currentRow * GBA_SCREEN_WIDTH + x + 40] = gba_ppu_colorToRgb(gba_ppu_vram_read16(0x06000000 | ((currentRow * 160 + x + offset) << 1)));
        }
    }
}

static inline void gba_ppu_drawLine() {
    gba_io_register_t *dispcnt = gba_io_getRegister(0x04000000);

    switch(dispcnt->value & 0x0007) {
        case 0: gba_ppu_drawMode0(); break;
        case 1: gba_ppu_drawMode1(); break;
        case 2: gba_ppu_drawMode2(); break;
        case 3: gba_ppu_drawMode3(); break;
        case 4: gba_ppu_drawMode4(); break;
        case 5: gba_ppu_drawMode5(); break;
        default: break;
    }
}

static inline void gba_ppu_onVblank() {
    gba_dma_onVblank();
    frontend_frame(gba_ppu_frameBuffer);
    gba_onFrame();
}

static inline void gba_ppu_onHblank() {
    gba_dma_onHblank();
}
